Implement new label format for large disks

pcd1193182 opened the pull request:
_Sponsored by: [Wasabi, Inc.; Klara, Inc.]_


### Motivation and Context
As disk sector sizes increase, we are able to store fewer and fewer uberblocks on a disk. This makes it increasingly difficult to recover from issues by rolling back to earlier TXGs. Eventually, sector sizes may become large enough that not even a single uberblock can be stored without having to do a partial write. In addition, new ZFS features often need space to store metadata (see, for example, the buffer used by RAIDZ expansion). This space is highly limited with the current disk layout.

### Description
This patch contains the logic for a new larger label format. This format is intended to support disks with large sector sizes. By using a larger label we can store more uberblocks and other critical pool metadata. We can also use the extra space to enable new features in ZFS going forwards. This initial commit does not add new capabilities, but provides the framework for them going forwards.

It also contains zdb and zhack support for the new label type, as well as tests that verify basic functionality of the new label. Currently, the size of the disk is used as a rubric for whether or not to enable the new label type, but that is open to change.

### How Has This Been Tested?
In addition to the tests added in this PR, I also ran the ZFS test suite with the tunable turned below the size of the disks in use. Some tests failed, but only for space estimation reasons, which could have been corrected with fixes to the tests. Similarly, I ran some ztest runs with the new label format.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Quality assurance (non-breaking change which makes the code more robust against bugs)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

pcd1193182 committed:
Implement new label format for large disks

This patch contains the logic for a new larger label format. This format
is intended to support disks with large sector sizes. By using a larger
label we can store more uberblocks and other critical pool metadata. We
can also use the extra space to enable new features in ZFS going
forwards. This initial commit does not add new capabilities, but
provides the framework for them going forwards.

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>
Sponsored-by: Wasabi, Inc.
Sponsored-by: Klara, Inc.

pcd1193182 committed:
Rob's feedback

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

pcd1193182 committed:
mav feedback

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

pcd1193182 commented on include/sys/vdev_impl.h:
```diff
@@ -532,6 +554,57 @@ typedef struct vdev_label {
+/* The size of the section that stores the pool config */
+#define	VDEV_TOC_POOL_CONFIG	"pool_config"
+/* The size of the section that stores auxilliary uberblocks */
+#define	VDEV_TOC_AUX_UBERBLOCK	"aux_uberblock"
```
AUX_UBERBLOCK was intended to be a label section that stored auxiliary uberblocks (like checkpoints & MMP). But this ultimately doesn't seem that useful; checkpoints don't need to store an uberblock anywhere in the label, because they can access the MOS (where the checkpointed uberblock is stored). The only way that would be especially useful is if the MOS was inaccessible due to corruption or a read-incompatible feature; but even then, with the number of uberblocks we can store now, we don't really need to store a special one in the label that's a copy of the checkpoint uberblock. That can be a future feature, if people want it.

robn reviewed (commented):
This seems to be a bit more polished version of the series I saw a few months ago, which I think bodes well - nothing bad or surprising seen since!

So what's still needed to move this forward? And what's the plan from here? Is the intent to get this merged and onto real pools before there's a new feature that requires it, or hold it until we need it? I'm guessing/hoping the former, to get operational experience and shake out any issues before we actually need it.

Any thoughts or guidance on how to use all this new space? I don't really at this stage, and I'm don't think there's a big long line of things waiting to use it. Regardless, as with most of our on-disk formats, if we're upgrading them to throw of limitations of the past, I would like this to be the last time we ever have to, and that in part means making sure we know how to use it and never break it!

pcd1193182 commented on tests/zfs-tests/tests/functional/large_label/large_label_001_pos.ksh:
```diff
@@ -0,0 +1,55 @@
+
+log_must create_pool -f $TESTPOOL "$DSK"0
+
+log_must zdb -l "$DSK"0
```
Yeah, if the test fails, it's nice to not have to modify and rerun just to see what the zdb output actually looked like, or if that was the command that failed directly, or we just failed to find the string we wanted.

pcd1193182 committed:
Try multi-ring setup for large labels

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

pcd1193182 committed:
Use three rings for storing Uberblocks

This provides a balance between frequent UBs at new txgs, and sparse
ones for historical purposes

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

pcd1193182 committed:
Add entry in man page

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

pcd1193182 committed:
tony feedback

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

pcd1193182 committed:
write all disks in parallel

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

pcd1193182 committed:
Add warning for sector size > 8k

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

tonyhutter reviewed (approved):
Approved, but please fix the minor typo in `vdev_impl.h`

pcd1193182 committed:
feedback

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

pcd1193182 committed:
alek's feedback

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

alek-p reviewed (commented):
Thanks for working on this Paul. Hope you don't mind that I'm reviewing this piecemeal

pcd1193182 committed:
more feedback

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

pcd1193182 committed:
Fix rebase onto master

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

alek-p commented on module/zfs/vdev_label.c:
seems like we're still assuming older label format here for the bootenv read

alek-p commented on cmd/zhack.c:
```diff
@@ -1221,9 +1478,18 @@ zhack_label_repair(const zhack_repair_op_t op, const int argc, char **argv)
+	if (large_label) {
+		for (int l = 0; l < VDEV_LABELS; l++) {
+			zhack_repair_one_label_large(op, fd,
+			    filesize, l, labels_repaired);
```
s/filesize/vdev_label_offset(filesize, l, 0, B_TRUE)/

alek-p commented on cmd/zhack.c:
```diff
@@ -1172,15 +1227,217 @@ zhack_repair_one_label(const zhack_repair_op_t op, const int fd,
 
 		zhack_repair_write_uberblock(vl,
 		    l, ashift, fd, byteswap, label_offset, labels_repaired);
+		if (large_label) {
```
should be `if (*large_label) {`

alek-p commented on cmd/zhack.c:
```diff
@@ -1172,15 +1227,217 @@ zhack_repair_one_label(const zhack_repair_op_t op, const int fd,
+
+	uint32_t bootenv_size, vc_size, sc_size;
+	uint32_t bootenv_offset, vc_offset, sc_offset;
+	if (!vdev_toc_get_secinfo(toc, VDEV_TOC_BOOT_REGION,
```
the presense of a TOC section should be optional right?

alek-p reviewed (commented):
thanks for working on this! Would be nice to see a large-label bootenv r/w test as a part of the patch

pcd1193182 committed:
Alek's feedback

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

pcd1193182 commented:
> thanks for working on this! Would be nice to see a large-label bootenv r/w test as a part of the patch

The main reason I haven't added one is that currently, there are no bootenv tests in the test suite in general. The bootenv stuff is only exposed through `libzfsbootenv`, which isn't tested as far as I can see. Do we want to add that test infrastructure as part of this PR?

alek-p commented:
> > thanks for working on this! Would be nice to see a large-label bootenv r/w test as a part of the patch
> 
> The main reason I haven't added one is that currently, there are no bootenv tests in the test suite in general. The bootenv stuff is only exposed through `libzfsbootenv`, which isn't tested as far as I can see. Do we want to add that test infrastructure as part of this PR?

yeah that's probably not directly related to this patch

alek-p commented on module/zfs/vdev_label.c:
```diff
@@ -1462,13 +1638,30 @@ vdev_label_write_bootenv(vdev_t *vd, nvlist_t *env)
 
-	bootenv = abd_borrow_buf_copy(abd, VDEV_PAD_SIZE);
+	uint32_t new_abd_size = (vd->vdev_large_label ?
+	    P2ROUNDUP(content_size + sizeof (uint64_t) + sizeof (zio_eck_t),
```
is it possible we never hit the VB_RAW nor VB_NVLIST switch cases above? If so then content_size would be uninitialized. If it's not possible let's initialize to zero and add an ASSERT(content_size != 0)?

pcd1193182 replied:
Well, in the default case we `return (EINVAL)`, so we always have to hit one of the two cases if we reach this code.

alek-p commented on module/zfs/vdev_label.c:
```diff
@@ -1491,20 +1684,82 @@ vdev_label_write_bootenv(vdev_t *vd, nvlist_t *env)
+			txg_kick(spa_get_dsl(spa),
+			    spa_syncing_txg(spa) + TXG_SIZE);
+			(void) spa_vdev_state_exit(spa, NULL, 0);
+			cv_wait(&vd->vdev_be_cv, &vd->vdev_be_lock);
```
we only broadcast vdev_be_cv on successful odd-label sync so failed syncs would leave this stuck?

alek-p commented on module/zfs/vdev_label.c:
```diff
@@ -1077,8 +1132,9 @@ vdev_aux_label_generate(vdev_t *vd, boolean_t reason_spare)
-int
-vdev_label_init(vdev_t *vd, uint64_t crtxg, vdev_labeltype_t reason)
+static int
+vdev_label_init_impl(vdev_t *vd, zio_t *pio, uint64_t crtxg,
```
looks like we don't end up using this added pio?

alek-p commented on man/man7/zpool-features.7:
```diff
@@ -757,6 +757,22 @@ are destroyed.
+the uberblock ring.
+However, because uberblocks always take at least one disk sector, large disk
+sectors result in unacceptably small uberblock rings.
+The new label format is 256MiB in size, with 128KiB reserved for the expanded
```
128 MiB?

alek-p commented on module/os/freebsd/zfs/vdev_label_os.c:
```diff
@@ -103,8 +103,9 @@ vdev_check_boot_reserve(spa_t *spa, vdev_t *childvd)
 	 */
 	zio_nowait(zio_vdev_child_io(pio, NULL, childvd,
-	    VDEV_BOOT_OFFSET - VDEV_LABEL_START_SIZE, abd, size, ZIO_TYPE_READ,
-	    ZIO_PRIORITY_ASYNC_READ, 0, vdev_child_done, pio));
```
uint64_t boot_off = childvd->vdev_large_label ?
      VDEV_RESERVE_OFFSET : VDEV_BOOT_OFFSET;
and use boot_off instead of VDEV_BOOT_OFFSET?

pcd1193182 replied:
Because this offset is written by external software, I'm not sure if it makes sense to update this code until that software is aware of the new label format.

alek-p commented on module/os/freebsd/zfs/vdev_label_os.c:
```diff
@@ -61,7 +61,7 @@ vdev_label_write_pad2(vdev_t *vd, const char *buf, size_t size)
 
 	zio = zio_root(spa, NULL, NULL, flags);
-	vdev_label_write(zio, vd, 0, pad2,
+	vdev_label_write(zio, vd, 0, B_FALSE, pad2,
```
seems like this should be large label aware too

alek-p commented on include/sys/vdev_impl.h:
```diff
@@ -474,24 +487,37 @@ struct vdev {
  */
 #define	MMP_BLOCKS_PER_LABEL	1
 
 /* The largest uberblock we support is 8k. */
```
looks like this comment is stale now

alek-p commented on include/sys/vdev_impl.h:
```diff
@@ -474,24 +487,37 @@ struct vdev {
 	(VDEV_UBERBLOCK_RING >> VDEV_UBERBLOCK_SHIFT(vd))
-#define	VDEV_UBERBLOCK_OFFSET(vd, n)	\
+#define	VDEV_UBERBLOCK_COUNT(vd)	\
+	(((vd)->vdev_large_label ? VDEV_LARGE_LABEL_SIZE - \
```
not sure i follow this. Shouldn't it just be `vdev_large_label ? VDEV_LARGE_UBERBLOCK_RING :  VDEV_UBERBLOCK_RING ...`

alek-p commented on module/zfs/spa.c:
```diff
@@ -1020,7 +1020,7 @@ spa_prop_set(spa_t *spa, nvlist_t *nvp)
 
 			/* Save time if the version is already set. */
-			if (ver == spa_version(spa))
+			if (spa_version(spa) >= ver)
```
why is this needed?

alek-p commented on module/zfs/vdev_label.c:
```diff
@@ -1923,12 +2248,88 @@ vdev_label_sync_ignore_done(zio_t *zio)
+	if (nvlist_pack(toc, &buf, &toc_buflen, NV_ENCODE_XDR, KM_SLEEP)) {
+		abd_free(toc_abd);
+		fnvlist_free(toc);
+		return;
```
the caller can't tell if this failed here which means config sync looks sucessful even though it wasn't

pcd1193182 replied:
The sync doesn't look successful because it returns, it looks successful because `vdev_label_sync_done` increments `good_writes`.

alek-p replied:
Although this is probably rare if not impossible since we assert the size and use KM_SLEEP - I would argue we either`zio->io_error = SET_ERROR(EIO)` here (simular to what we do in `vdev_label_sync_top_done()`) or track how many `large_good_writes` we've had (>=1 hopefully).

Otherwise, I think, if pack fails and we skip the large-label writes, the sync still looks fine if the legacy `vdev_label_writes()` succeed and bumps `good_writes`.

alek-p commented on module/zfs/vdev.c:
```diff
@@ -2287,22 +2310,40 @@ vdev_open(vdev_t *vd)
-
+	boolean_t large_label;
+	if (spa_version(spa) < SPA_VERSION_FEATURES ||
+	    !(spa_feature_is_enabled(spa, SPA_FEATURE_LARGE_LABEL) ||
```
vdev_impl.h says "l2arc and spare devices are excluded" but that doesn't seem to be enforced here

pcd1193182 committed:
address feedback

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>

alek-p commented on module/zfs/vdev_label.c:
```diff
@@ -1728,6 +2003,12 @@ vdev_uberblock_load(vdev_t *rvd, uberblock_t *ub, nvlist_t **config)
+		vdev_dbgmsg(rvd, "failed to read label config. "
+		    "Trying again without full uberblock ring.");
+		try_hard = B_TRUE;
+		goto retry;
```
need to `spa_config_exit(spa, SCL_ALL, FTAG)` before the retry goto

alek-p reviewed (commented):
This looks really close. I imagine this format will take over the legacy labels pretty fast.
I'm certianly interested in being able to have more recovery opportunities.

pcd1193182 committed:
Alek's feedback

Signed-off-by: Paul Dagnelie <paul.dagnelie@klarasystems.com>
